src/orchestrator.ts

```diff
@@ -18,93 +18,85 @@ type OrchestratorOptions = {
-      fork(function*() {
-        yield receive(orchestrator, { ready: "proxy" });
-      });
-      fork(function*() {
-        yield receive(orchestrator, { ready: "command" });
-      });
-      fork(function*() {
-        yield receive(orchestrator, { ready: "connection" });
-      });
-      fork(function*() {
-        yield receive(orchestrator, { ready: "agent" });
-      });
-      fork(function*() {
-        yield receive(orchestrator, { ready: "app" });
-      });
-      fork(function*() {
-        yield receive(orchestrator, { ready: "test-files" });
-      });
-    });
-
-    console.log("[orchestrator] running!");
-
-    if(options.delegate) {
-      options.delegate.send({ ready: "orchestrator" });
-    }
+export function* createOrchestrator(options: OrchestratorOptions): Operation {
+  let orchestrator = yield ({ resume, context: { parent }}) => resume(parent);
+
+  console.log('[orchestrator] starting');
+
+  yield fork(createProxyServer(orchestrator, {
+    port: options.proxyPort,
+    targetPort: options.appPort,
+    inject: `<script src="http://localhost:${options.agentPort}/harness.js"></script>`,
+  }));
+
+  yield fork(createCommandServer(orchestrator, {
+    port: options.commandPort,
+  }));
+
+  yield fork(createConnectionServer(orchestrator, {
+    port: options.connectionPort,
+    proxyPort: options.proxyPort,
+    testFilePort: options.testFilePort,
+  }));
+
+  yield fork(createAgentServer(orchestrator, {
+    port: options.agentPort,
+  }));
+
+  yield fork(createAppServer(orchestrator, {
+    dir: options.appDir,
+    command: options.appCommand,
+    args: options.appArgs,
+    env: options.appEnv,
+    port: options.appPort,
+  }));
+
+  yield fork(createTestFileWatcher(orchestrator, {
+    files: options.testFiles,
+    manifestPath: options.testManifestPath,
+  }));
+
+  // wait for manifest before starting test file server
+  yield receive({ ready: "manifest" }, orchestrator);
+
+  yield fork(createTestFileServer(orchestrator, {
+    files: options.testFiles,
+    manifestPath: options.testManifestPath,
+    port: options.testFilePort,
+  }));
+
+  yield all([
+    receive({ ready: "proxy" }, orchestrator),
+    receive({ ready: "command" }, orchestrator),
+    receive({ ready: "connection" }, orchestrator),
+    receive({ ready: "agent" }, orchestrator),
+    receive({ ready: "app" }, orchestrator),
+    receive({ ready: "test-files" }, orchestrator)
+  ]);
+
+  console.log("[orchestrator] running!");
+
+  if(options.delegate) {
+    yield send({ ready: "orchestrator" }, options.delegate);
+  }
 
-    try {
-      while(true) {
-        yield receive(orchestrator);
-      }
-    } finally {
-      console.log("[orchestrator] shutting down!");
+  try {
+    while(true) {
+      yield receive(orchestrator);
     }
+  } finally {
+    console.log("[orchestrator] shutting down!");
+  }
+}
+
+function* all(operations: Operation[]): Operation {
+  for (let operation of operations) {
+    yield fork(operation);
```

Comment:
The behavior I was going for was to have all operations run in parallel, similar to `Promise.all`, hence the `fork`.

Reply:
In this case though, it doesn't really matter if they execute in serial or parallel.

Reply:
But in this case, I think that running in serial or parallel would have the exact same effect.

Reply:
It's more that we need to initiate the operations, and then actually block and wait for all of them to complete. As far as I can see, this only initiates the operations, never actually blocking to wait for them to complete. I could be wrong though?

Reply:
Because of the strict guarantees of structured concurrency, the `all` operation itself cannot be complete until all its child forks are also complete, so while `all` will almost immediately transition into the `waiting` state, it will not transition into `completed` until it has no more blocking children.